File: backend/app/schemas/dhcp.py
"""DHCP schemas — pools, servers, networks, leases (P21 Stage 10 / #1)."""

from __future__ import annotations

from pydantic import BaseModel, Field


# ---------------- Pools ----------------


class DhcpPoolPublic(BaseModel):
    id: str | None
    name: str
    ranges: str
    next_pool: str | None
    comment: str | None


class DhcpPoolCreate(BaseModel):
    name: str = Field(min_length=1, max_length=64)
    ranges: str = Field(min_length=1, max_length=512)
    next_pool: str | None = Field(default=None, max_length=64)
    comment: str | None = Field(default=None, max_length=255)


class DhcpPoolUpdate(BaseModel):
    name: str | None = Field(default=None, max_length=64)
    ranges: str | None = Field(default=None, max_length=512)
    next_pool: str | None = Field(default=None, max_length=64)
    comment: str | None = Field(default=None, max_length=255)


# ---------------- Servers ----------------


class DhcpServerPublic(BaseModel):
    id: str | None
    name: str
    interface: str
    address_pool: str | None
    lease_time: str | None
    authoritative: str | None
    disabled: bool
    comment: str | None


class DhcpServerCreate(BaseModel):
    name: str = Field(min_length=1, max_length=64)
    interface: str = Field(min_length=1, max_length=64)
    address_pool: str | None = Field(default=None, max_length=64)
    lease_time: str | None = Field(default=None, max_length=32)
    authoritative: str | None = Field(default=None, max_length=32)
    disabled: bool = False
    comment: str | None = Field(default=None, max_length=255)


class DhcpServerUpdate(BaseModel):
    name: str | None = Field(default=None, max_length=64)
    interface: str | None = Field(default=None, max_length=64)
    address_pool: str | None = Field(default=None, max_length=64)
    lease_time: str | None = Field(default=None, max_length=32)
    authoritative: str | None = Field(default=None, max_length=32)
    disabled: bool | None = None
    comment: str | None = Field(default=None, max_length=255)


# ---------------- Networks ----------------


class DhcpNetworkPublic(BaseModel):
    id: str | None
    address: str
    gateway: str | None
    netmask: str | None
    dns_servers: str | None
    ntp_servers: str | None
    domain: str | None
    comment: str | None


class DhcpNetworkCreate(BaseModel):
    address: str = Field(min_length=1, max_length=64)
    gateway: str | None = Field(default=None, max_length=64)
    netmask: str | None = Field(default=None, max_length=32)
    dns_servers: str | None = Field(default=None, max_length=255)
    ntp_servers: str | None = Field(default=None, max_length=255)
    domain: str | None = Field(default=None, max_length=255)
    comment: str | None = Field(default=None, max_length=255)


class DhcpNetworkUpdate(BaseModel):
    address: str | None = Field(default=None, max_length=64)
    gateway: str | None = Field(default=None, max_length=64)
    netmask: str | None = Field(default=None, max_length=32)
    dns_servers: str | None = Field(default=None, max_length=255)
    ntp_servers: str | None = Field(default=None, max_length=255)
    domain: str | None = Field(default=None, max_length=255)
    comment: str | None = Field(default=None, max_length=255)


# ---------------- Leases ----------------


class DhcpLeasePublic(BaseModel):
    id: str | None
    address: str
    mac_address: str
    host_name: str | None
    client_id: str | None
    status: str | None
    server: str | None
    expires_at_iso: str | None
    dynamic: bool = False
    blocked: bool = False
    comment: str | None


class DhcpLeaseCommentUpdate(BaseModel):
    comment: str | None = Field(default=None, max_length=255)

File: frontend/src/lib/dhcp.ts
import { api, readErrorMessage } from "@/lib/api";

// ---------------- Pools ----------------

export interface DhcpPool {
  id: string | null;
  name: string;
  ranges: string;
  next_pool: string | null;
  comment: string | null;
}

export interface DhcpPoolCreate {
  name: string;
  ranges: string;
  next_pool?: string | null;
  comment?: string | null;
}

export type DhcpPoolUpdate = Partial<DhcpPoolCreate>;

export async function listDhcpPools(deviceId: string): Promise<DhcpPool[]> {
  return api.get(`devices/${deviceId}/dhcp/pools`).json<DhcpPool[]>();
}

export async function createDhcpPool(
  deviceId: string,
  payload: DhcpPoolCreate,
): Promise<{ id: string }> {
  try {
    return await api
      .post(`devices/${deviceId}/dhcp/pools`, { json: payload })
      .json<{ id: string }>();
  } catch (e) {
    throw new Error(await readErrorMessage(e));
  }
}

export async function updateDhcpPool(
  deviceId: string,
  poolId: string,
  payload: DhcpPoolUpdate,
): Promise<void> {
  try {
    await api.patch(`devices/${deviceId}/dhcp/pools/${encodeURIComponent(poolId)}`, {
      json: payload,
    });
  } catch (e) {
    throw new Error(await readErrorMessage(e));
  }
}

export async function deleteDhcpPool(deviceId: string, poolId: string): Promise<void> {
  try {
    await api.delete(`devices/${deviceId}/dhcp/pools/${encodeURIComponent(poolId)}`);
  } catch (e) {
    throw new Error(await readErrorMessage(e));
  }
}

// ---------------- Servers ----------------

export interface DhcpServer {
  id: string | null;
  name: string;
  interface: string;
  address_pool: string | null;
  lease_time: string | null;
  authoritative: string | null;
  disabled: boolean;
  comment: string | null;
}

export interface DhcpServerCreate {
  name: string;
  interface: string;
  address_pool?: string | null;
  lease_time?: string | null;
  authoritative?: string | null;
  disabled?: boolean;
  comment?: string | null;
}

export type DhcpServerUpdate = Partial<DhcpServerCreate>;

export async function listDhcpServers(deviceId: string): Promise<DhcpServer[]> {
  return api.get(`devices/${deviceId}/dhcp/servers`).json<DhcpServer[]>();
}

export async function createDhcpServer(
  deviceId: string,
  payload: DhcpServerCreate,
): Promise<{ id: string }> {
  try {
    return await api
      .post(`devices/${deviceId}/dhcp/servers`, { json: payload })
      .json<{ id: string }>();
  } catch (e) {
    throw new Error(await readErrorMessage(e));
  }
}

export async function updateDhcpServer(
  deviceId: string,
  serverId: string,
  payload: DhcpServerUpdate,
): Promise<void> {
  try {
    await api.patch(
      `devices/${deviceId}/dhcp/servers/${encodeURIComponent(serverId)}`,
      { json: payload },
    );
  } catch (e) {
    throw new Error(await readErrorMessage(e));
  }
}

export async function deleteDhcpServer(deviceId: string, serverId: string): Promise<void> {
  try {
    await api.delete(`devices/${deviceId}/dhcp/servers/${encodeURIComponent(serverId)}`);
  } catch (e) {
    throw new Error(await readErrorMessage(e));
  }
}

// ---------------- Networks ----------------

export interface DhcpNetwork {
  id: string | null;
  address: string;
  gateway: string | null;
  netmask: string | null;
  dns_servers: string | null;
  ntp_servers: string | null;
  domain: string | null;
  comment: string | null;
}

export interface DhcpNetworkCreate {
  address: string;
  gateway?: string | null;
  netmask?: string | null;
  dns_servers?: string | null;
  ntp_servers?: string | null;
  domain?: string | null;
  comment?: string | null;
}

export type DhcpNetworkUpdate = Partial<DhcpNetworkCreate>;

export async function listDhcpNetworks(deviceId: string): Promise<DhcpNetwork[]> {
  return api.get(`devices/${deviceId}/dhcp/networks`).json<DhcpNetwork[]>();
}

export async function createDhcpNetwork(
  deviceId: string,
  payload: DhcpNetworkCreate,
): Promise<{ id: string }> {
  try {
    return await api
      .post(`devices/${deviceId}/dhcp/networks`, { json: payload })
      .json<{ id: string }>();
  } catch (e) {
    throw new Error(await readErrorMessage(e));
  }
}

export async function updateDhcpNetwork(
  deviceId: string,
  networkId: string,
  payload: DhcpNetworkUpdate,
): Promise<void> {
  try {
    await api.patch(
      `devices/${deviceId}/dhcp/networks/${encodeURIComponent(networkId)}`,
      { json: payload },
    );
  } catch (e) {
    throw new Error(await readErrorMessage(e));
  }
}

export async function deleteDhcpNetwork(
  deviceId: string,
  networkId: string,
): Promise<void> {
  try {
    await api.delete(`devices/${deviceId}/dhcp/networks/${encodeURIComponent(networkId)}`);
  } catch (e) {
    throw new Error(await readErrorMessage(e));
  }
}

// ---------------- Leases ----------------

export interface DhcpLease {
  id: string | null;
  address: string;
  mac_address: string;
  host_name: string | null;
  client_id: string | null;
  status: string | null;
  server: string | null;
  expires_at_iso: string | null;
  dynamic: boolean;
  blocked: boolean;
  comment: string | null;
}

export async function listDhcpLeases(deviceId: string): Promise<DhcpLease[]> {
  return api.get(`devices/${deviceId}/dhcp/leases`).json<DhcpLease[]>();
}

export async function makeLeaseStatic(
  deviceId: string,
  leaseId: string,
): Promise<void> {
  try {
    await api.post(
      `devices/${deviceId}/dhcp/leases/${encodeURIComponent(leaseId)}/make-static`,
    );
  } catch (e) {
    throw new Error(await readErrorMessage(e));
  }
}

export async function setLeaseComment(
  deviceId: string,
  leaseId: string,
  comment: string | null,
): Promise<void> {
  try {
    await api.patch(
      `devices/${deviceId}/dhcp/leases/${encodeURIComponent(leaseId)}/comment`,
      { json: { comment } },
    );
  } catch (e) {
    throw new Error(await readErrorMessage(e));
  }
}

export async function deleteLease(
  deviceId: string,
  leaseId: string,
): Promise<void> {
  try {
    await api.delete(`devices/${deviceId}/dhcp/leases/${encodeURIComponent(leaseId)}`);
  } catch (e) {
    throw new Error(await readErrorMessage(e));
  }
}
